### DISCUSSION TITLE
Scripts

### DISCUSSION THREAD

**Discussion by xcasas** [ImSwitch/ImSwitch#95] [General]
I create this discussion to upload scripts

**Comment by xcasas** [comment]
**How to turn on a laser, move a positioner and take a snap**
```
import time

mainWindow.setCurrentModule('imcontrol')

laserName = api.imcontrol.getLaserNames()[2]
positionerXName = api.imcontrol.getPositionerNames()[0]
positionerYName = api.imcontrol.getPositionerNames()[1]

api.imcontrol.setLaserActive(laserName, True)
api.imcontrol.movePositioner(positionerXName, "X", 3)
time.sleep(1)
api.imcontrol.snapImage() 

```

**Comment by xcasas** [comment]
**Automated photophysics characterization, running a scan for different laser powers**
```
import numpy as np
mainWindow.setCurrentModule('imcontrol')

values_405 = np.array([0, 0, 0, 0, 5, 10, 25, 50, 100, 200])
lasername_405 = api.imcontrol.getLaserNames()[0]

# Start Recording
api.imcontrol.startRecording()
for i in values_405:
	# Set Laser Value
	api.imcontrol.changeScanPower(lasername_405, i)
	# Run Scan
	api.imcontrol.runScan()
	# Wait for finish
	waitForScanningToEnd = getWaitForSignal(api.imcontrol.signals().scanEnded)
	waitForScanningToEnd()
	
api.imcontrol.stopRecording()
```

**Comment by beniroquai** [comment]
I found that you can export a list with all the API functions here:
```py
    return pythontools.dictToROClass(exportedFuncs,
                                     missingAttributeErrorMsg=missingAttributeErrorMsg)
```
Is there any other way to do it in a more easy way? A bit complicated doing that all the time - perhaps having a document auto-generated from here with all the python stubs?

**Reply by xcasas** [reply]
Now they are already automatically exported, you just have to add the decorator @APIExport before the function, I am not sure what you want to do sorry? You mean that they are exported without having to add the decorator?

**Reply by beniroquai** [reply]
Maybe just a doucment that lists all available functions that can be accessed from 3rd party scripts. I would need to have them if I'm about to wrap it e.g. for the Pyro client.

**Reply by xcasas** [reply]
[this script](https://github.com/kasasxav/ImSwitch/blob/9ba8d5d0ec243fe9c401fd4308ea90d770b45fb0/tools/generateapidocs.py) exports the functions [to this document](https://github.com/kasasxav/ImSwitch/blob/master/docs/api/api.imcontrol.rst) which is part of the [readthedocs](https://imswitch.readthedocs.io/en/stable/api/api.imcontrol.html)

**Reply by beniroquai** [reply]
Exactly what I wanted! :-)
